src/Components/WebView/WebView/src/Services/WebViewNavigationManager.cs

```diff
@@ -20,8 +28,56 @@ public void LocationUpdated(string newUrl, string? state, bool intercepted)
         NotifyLocationChanged(intercepted);
     }
 
+    public void HandleLocationChanging(int callId, string uri, bool intercepted)
+    {
+        NotifyLocationChanging(uri, intercepted, result =>
+        {
+            bool success;
+
+            if (result.Exception is { } exception)
+            {
+                success = false;
+                Log.NavigationFailed(_logger, uri, exception);
+            }
+            else if (result.Canceled)
+            {
+                success = false;
+                Log.NavigationCanceled(_logger, uri);
+            }
+            else
+            {
+                success = true;
+            }
+
+            _ipcSender.EndLocationChanging(callId, success);
+        });
+    }
+
     protected override void NavigateToCore(string uri, NavigationOptions options)
     {
-        _ipcSender.Navigate(uri, options);
+        NotifyLocationChanging(uri, false, result =>
+        {
+            if (result.Exception is { } exception)
+            {
+                Log.NavigationFailed(_logger, uri, exception);
+            }
+            else if (result.Canceled)
+            {
+                Log.NavigationCanceled(_logger, uri);
+            }
+            else
+            {
+                _ipcSender.Navigate(uri, options);
+            }
+        });
+    }
```

Comment:
Possibly, but they couldn't know right away since the navigation isn't guaranteed to complete by the time `NavigateTo` returns. Were you thinking we could add some other event to `NavigationManager` like `NavigationCanceled`? If so, I think that should be something we add as a follow-up to these changes.

Reply:
I was thinking we could add `ValueTask<bool> NavigateToAsync` which lets you know if the navigation is cancelled (or succeeds in this case), and NavigateTo which you use if you just don't care. 

If you navigate programmatically but that does nothing, I think you want to be able to react in code. Specially since the navigation might take some time to take effect.

Reply:
Follow up:

@javiercn and I were discussing this and trying to decide if it's important to have in 7.0 or whether it would be OK to leave until we have a more concrete requirement from developer scenarios. We didn't get a definite conclusion.

I've been trying to think of the most plausible scenario for some code genuinely needing to know whether the navigation was cancelled or not, and it's quite hard to think of one. The closest I've got is some flow where an action button becomes disabled once you press it, like this:

```cs
private async Task SaveChangesAsync() // Invoked via a button click
{
    this.disableSaveButton = true;
    if (await myData.SaveAsync())
    {
        // Save succeeded; redirect away
       NavManager.NavigateTo("list");
    }
    else
    {
        // Save failed - re-enable the button
        this.disableSaveButton = false;
    }
}
```

If the `NavigateTo` call gets cancelled, the user will be stuck because the button isn't re-enabled.

However I don't think this is a 100% valid scenario as presented here, because

1. There's still the question of *why would some completely unrelated code be interfering with navigation*? Normally it only makes sense for the code that manages the edit flow to block the navigation.
2. Even if you did re-enable the save button, that's probably not what you want to do (they have already saved the changes by now). You probably want to retry the navigation or something, but it's very unclear what sort of user gesture would trigger that. TBH the entire UX flow doesn't really make sense if the navigation might randomly not happen.

So I'm still on the lookout for a convincing account of a realistic UI that would make use of the `NavigateToAsync` result value. Maybe there's a simple example someone can come up with!